helper/winsvc/windows_service_windows.go

```diff
@@ -0,0 +1,230 @@
+}
+
+func (s *windowsService) Configure(config WindowsServiceConfiguration) error {
+	serviceCfg, err := s.service.Config()
+	if err != nil {
+		return err
+	}
+
+	serviceCfg.StartType = uint32(config.StartType)
+	serviceCfg.DisplayName = config.DisplayName
+	serviceCfg.Description = config.Description
+	serviceCfg.BinaryPathName = config.BinaryPathName
+
+	if err := s.service.UpdateConfig(serviceCfg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *windowsService) Start() error {
+	if running, _ := s.IsRunning(); running {
+		return nil
+	}
+
+	if err := s.service.Start(); err != nil {
+		return err
+	}
+
+	if err := s.waitFor(s.IsRunning); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *windowsService) Stop() error {
+	if stopped, _ := s.IsStopped(); stopped {
+		return nil
+	}
+
+	if _, err := s.service.Control(svc.Stop); err != nil {
+		return err
+	}
+
+	if err := s.waitFor(s.IsStopped); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *windowsService) Close() error {
+	return s.service.Close()
+}
+
+func (s *windowsService) Delete() error {
+	return s.service.Delete()
+}
+
+func (s *windowsService) IsRunning() (bool, error) {
+	return s.isService(svc.Running)
+}
+
+func (s *windowsService) IsStopped() (bool, error) {
+	return s.isService(svc.Stopped)
+}
+
+func (s *windowsService) EnableEventlog() error {
+	// Check if the service is already setup in the eventlog
+	key, err := registry.OpenKey(registry.LOCAL_MACHINE,
+		EVENTLOG_REGISTRY_PATH+`\`+s.Name(),
+		registry.ALL_ACCESS,
+	)
+	defer key.Close()
+
+	// If it could not be opened, assume error is caused
+	// due to nonexistence. If it was for some other reason
+	// the error will be encountered again when attempting to
+	// create.
+	if err != nil {
+		if err := eventlog.InstallAsEventCreate(s.Name(), EVENTLOG_SUPPORTED_EVENTS); err != nil {
+			return err
+		}
+	} else {
+		// Since the service is already registered, just
+		// ensure it is properly configured. Currently
+		// that just means the supported events.
+		val, _, err := key.GetIntegerValue(EVENTLOG_SUPPORTED_EVENTS_KEY)
+		if err != nil || uint32(val) != EVENTLOG_SUPPORTED_EVENTS {
+			if err := key.SetDWordValue(EVENTLOG_SUPPORTED_EVENTS_KEY, EVENTLOG_SUPPORTED_EVENTS); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *windowsService) DisableEventlog() error {
+	// Check if the service is currently enabled in the eventlog
+	key, err := registry.OpenKey(registry.LOCAL_MACHINE,
+		EVENTLOG_REGISTRY_PATH+`\`+s.Name(),
+		registry.READ,
+	)
+	if errors.Is(err, fs.ErrNotExist) {
+		return nil
+	}
+	defer key.Close()
+
+	return eventlog.Remove(s.Name())
+}
+
+func (s *windowsService) isService(state svc.State) (bool, error) {
+	status, err := s.service.Query()
+	if err != nil {
+		return false, err
+	}
+
+	return status.State == state, nil
+}
+
+func (s *windowsService) waitFor(condition func() (bool, error)) error {
+	for range WINDOWS_SERVICE_STATE_TIMEOUT * 4 {
```

Comment:
what happens if someone hits ctrl+c ? (is that windows' interrupt mechanism, too?)

or am I being too linux-brained right now wondering about interrupting something within these 60*4/4 seconds?

I suppose I'm surprised that there's no context or channel business going on around here.